torch/csrc/jit/passes/loop_unrolling.cpp

```diff
@@ -0,0 +1,217 @@
+#include "torch/csrc/jit/interned_strings.h"
+#include "torch/csrc/jit/symbolic_variable.h"
+#include "torch/csrc/jit/passes/dead_code_elimination.h"
+
+namespace torch { namespace jit {
+
+namespace {
+
+static constexpr int64_t kUnrollFactor = 8;
+static constexpr int64_t kMaxBodySize = 16;
+static constexpr int64_t kMaxBodyRepeats = 64;
+
+bool isTrueConstant(Value *val) {
+  Node *producer = val->node();
+  if (producer->kind() != prim::Constant)
+    return false;
+  auto value = producer->t(attr::value);
+  return value.type() == at::CPU(at::kByte) && value.dim() == 0 && value.toCLong() == 1;
+}
+
+bool isForLoop(Node* node) {
+  if (node->kind() != prim::Loop)
+    return false;
+  Value *start_cond = node->inputs().at(1);
+  Value *continue_cond = node->blocks().at(0)->outputs().at(0);
+  return isTrueConstant(start_cond) && isTrueConstant(continue_cond);
+}
+
+int64_t limitedBlockSize(Block *body, int64_t limit) {
+  auto it = body->nodes().begin();
+  auto end = body->nodes().end();
+  for (int64_t i = 0; i < limit; ++i, ++it) {
+    for (Block *subblock : it->blocks()) {
+      i += limitedBlockSize(subblock, limit);
+    }
+    if (it == end) {
+      return i;
+    }
+  }
+  return limit;
+}
+
+bool isSmallBlock(Block *body) {
+  return limitedBlockSize(body, kMaxBodySize + 1) <= kMaxBodySize;
+}
+
+at::optional<int64_t> getConstantLength(Node *loop) {
+  Value *trip_count = loop->inputs().at(0);
+  if (trip_count->node()->kind() != prim::Constant)
+    return at::nullopt;
+  return {trip_count->node()->t(attr::value).toCLong()};
+}
+
+// XXX: This function can only be called with a loop that is guaranteed to execute EXACTLY ONCE.
+void inlineBody(Node *loop) {
+  auto graph = loop->owningGraph();
+  auto body = loop->blocks().at(0);
+  WithInsertPoint insert_point_guard { loop };
+
+  std::unordered_map<Value*, Value*> value_map;
+  auto get_value = [&](Value *v) {
+    auto it = value_map.find(v);
+    if (it != value_map.end())
+      return it->second;
+    return v;
+  };
+
+  // Loop node has extra (max_iters, initial_cond) inputs,
+  // body has an extra (loop_counter) input.
+  for (size_t i = 2; i < loop->inputs().size(); ++i) {
+    value_map[body->inputs()[i - 1]] = loop->inputs()[i];
+  }
+
+  for (Node *orig : body->nodes()) {
+    Node *clone = graph->insertNode(graph->createClone(orig, get_value));
+    for (size_t i = 0; i < orig->outputs().size(); ++i) {
+      value_map[orig->outputs()[i]] = clone->outputs()[i];
+    }
+  }
+  for (size_t i = 0; i < loop->outputs().size(); ++i) {
+    loop->outputs().at(i)->replaceAllUsesWith(get_value(body->outputs().at(i + 1)));
+  }
+  // XXX: it is extremely important to destroy the loop in here. DCE might not be able
+  // to conclude that it's safe, because the loop might contain side effects.
+  loop->destroy();
+}
+
+void repeatBody(Block *body, int64_t times) {
+  // We will be adding nodes to the body, so cache the initial start and end.
+  // XXX: they are both inclusive
```

Comment:
It's because if the `body_end` pointed to the `return_node`, then it would be moving further and further away from the beginning after every node is cloned, and the loop wouldn't terminate. I'll add a comment.